src/opers.c

```diff
@@ -1821,2769 +1786,615 @@ static inline Obj CacheOper (
-
-    do {
-      /* The next line depends on the implementation of INTOBJS */
-      prec = (Obj)(((Int)prec) +4);
-      method = 0;
-
-      /* recalculate cache each pass, in case of GC     */
-      cache = 1+ADDR_OBJ( CacheOper( oper, 1 ) );
 
-      /* Up to CACHE_SIZE methods might be in the cache */
-      if (prec < INTOBJ_INT(CACHE_SIZE))
-        {
-          for (i = 0;  i < 3*CACHE_SIZE; i+= 3) {
-            if (  cache[i+1] == prec && cache[i+2] == id1 ) {
-              method = cache[i];
+static inline Obj __attribute__((always_inline))
+GetMethodCached(Obj oper, UInt n, Obj prec, Obj ids[])
+{
+    UInt  typematch;
+    Obj * cache;
+    Obj   method = 0;
+    UInt  i;
+
+    cache = 1 + ADDR_OBJ(CacheOper(oper, n));
+
+    /* Up to CACHE_SIZE methods might be in the cache */
+    if (prec < INTOBJ_INT(CACHE_SIZE)) {
+        /* This loop runs through those */
+        UInt target = (n + 2) * INT_INTOBJ(prec);
+        for (i = target; i < (n + 2) * CACHE_SIZE; i += n + 2) {
+            if (cache[i + 1] == prec) {
+                typematch = 1;
+                /* This loop runs over the arguments, should be compiled away
+                 */
+                for (UInt j = 0; j < n; j++) {
+                    if (cache[i + j + 2] != ids[j]) {
+                        typematch = 0;
+                        break;
+                    }
+                }
+                if (typematch) {
+                    method = cache[i];
 #ifdef COUNT_OPERS
-              OperationHit++;
+                    CacheHitStatistics[INT_INTOBJ(prec)][i / (n + 2)][n]++;
 #endif
-              break;
-            }
-          }
-        }
-      
-      /* otherwise try to find one in the list of methods                    */
-      if (!method)
-        {
-          if (prec == INTOBJ_INT(0))
-            method = CALL_2ARGS( Method1Args, oper, type1 );
-          else
-            method = CALL_3ARGS( NextMethod1Args, oper, prec, type1 );
-          
-          /* If there was no method found, then pass the information needed for
-             the error reporting. This function rarely returns */
-          if (method == Fail)
-          {
-            Obj args[1];
-            args[0] = arg1;
-            while (method == Fail)
-              method = CallHandleMethodNotFound( oper, 1, (Obj *) args, 0, 0, prec);
-          }
-          /* update the cache */
-          if (method && prec < INTOBJ_INT(CACHE_SIZE))
-            {
-              Bag cacheBag = GET_METHOD_CACHE( oper, 1 );
-              cache = 1+ADDR_OBJ( cacheBag );
-              cache[3*STATE(CacheIndex)] = method;
-              cache[3*STATE(CacheIndex)+1] = prec;
-              cache[3*STATE(CacheIndex)+2] = id1;
-              STATE(CacheIndex) = (STATE(CacheIndex) + 1) % CACHE_SIZE;
-              CHANGED_BAG( cacheBag );
+                    if (i > target) {
+                        Obj buf[n + 2];
+                        memmove((void *)buf, (void *)(cache + i),
```

Comment:
Why use `memmove` here? Since `buf` is allocated above, we know it cannot overlap with `cache`, hence `memcpy` is safe to use, and potentially very slightly faster.